Oversized bag tests

paul-rl opened the pull request:
## Description
Initial implementation of tests corresponding to the oversized bag data structure.

## Changes in the codebase
Created some tests for a variety of functions pertaining to the oversized bag. Mainly, created tests for the following features:

- Basic tagging #23 
- Rank 0 async_insert #24 
- for_all #27 
- for_all (pair) #28 
- for_all (split pair) #29 

Additionally, created skeletons for the following tests that will be getting implemented at a later point in time:
- rebalance #30 
- rebalance with non-standard rebalance sizes #31 

Something to note as well is that the shuffle functions (#26) are not implemented for now.

## Additional Information
A lot of the tests outlined here are very similar to the original bag tests. I don't expect everything here to be correct on the first try, and would appreciate any input on how it should be changed.

paul-rl7 committed:
Added basic test shells

paul-rl7 committed:
Added majority of test functions for initial review

Daniel-DeArmas commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,123 @@
+  }
+
+  //
+  // Test for_all (pair)
```
I noticed for the reference file "test/test_bag.cpp", the "Test for_all (pair)" is commented out. What is the reason it is not commented here

Daniel-DeArmas commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,123 @@
+  }
+
+  //
+  // Test rebalance
```
The rebalance functions are empty here but functional in the bag test file. Wouldn't they work the same way?

Daniel-DeArmas commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,123 @@
+
+  // Test basic tagging
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
```
This is more of a nitpick but I personally believe bag(world) should have its name changed to something like obag(world) or overbag(world) to clarify it is using the oversized bag and not a regular bag.

Daniel-DeArmas reviewed (approved):
This is a good starting point for the test suite

paul-rl7 committed:
Merge branch 'feature-oversized-bag-dev' into oversized_bag_tests

paul-rl7 committed:
Fixed comment for shuffle after Ryan discussed with Roger

paul-rl7 committed:
Added oversized bag tests to CMakeLists.txt

paul-rl7 committed:
Added changes to local_count to fix const issue ty ryan

paul-rl7 committed:
Isolated rank0 async insert test

ryan-dozier committed:
fixed local_count

paul-rl7 committed:
Added other tests commented out to keep rank 0 isolated to troubleshoot local_count

Daniel-DeArmas committed:
Added Rebalance Tests(issues)

Added rebalance tests to oversized_bag_test but I ran into problems.  I attempted changing the input data and the clear function but was unable to fix the issue. The test file gets stuck in a loop when the rebalance tests are not commented out. This may be a problem with the rebalance function or the setup of the tests.

paul-rl7 committed:
Uncommented complete oversized bag tests

paul-rl7 committed:
Fixed merge conflict caused by Danny's commit

paul-rl7 committed:
Merge branch 'feature-oversized-bag-dev' into oversized_bag_tests

paul-rl commented:
After these changes were made, there was some additional work done by @Daniel-DeArmas in ensuring that the re-balance functions were implemented correctly alongside with some fixes by @ryan-dozier to ensure that the `load_existing_files()` function implemented by Danny

paul-rl commented:
The combination of all of the changes made in this branch will do the following:

- closes #23 
- closes #24 
- closes #25
- closes #27 
- closes #28 
- closes #29 

Something to note here is that issue #26 is considered out of scope due to the local and global shuffle functions themselves being considered out of scope.

Additionally, the tests do not currently pass as the oversized_bag implementation is not yet complete.

paul-rl7 committed:
Merged changes from feature branch

paul-rl7 committed:
Uncommitted some code in file overflow test, included all ranks async_insert test

paul-rl7 committed:
Uncommented rebalancing tests as infinite loop issue is no longer occurring

paul-rl7 committed:
Added some fixes missed by Ryan's fixing of the persist state

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+      std::set<std::string> all_data;
+      bbag.gather(all_data, 0);
+      if (world.rank0()) {
+        YGM_ASSERT_RELEASE(all_data.size() == 3);
```
I don't believe the size of all_data will be 3 here. You seem to test the gather earlier. For the moment I think its good to test gather. HOWEVER, I am a bit concerned what a all_gather operation even means in the context of a bag with potentially millions/billions of entries.

paul-rl replied:
Added a bbag.clear() to ensure this test case is isolated and files persisting from previous tests are deleted.

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+  //
+  // Test Rebalance
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
```
This almost certainly needs a bbag.clear(), because the constructor persists previous files the size will be non-zero on startup.

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+  // 
+  // Test for_all
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
```
Same as above, the previous state will be persisted even across creating a 'new' local oversized bag

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+    bbag.for_all([&count](std::string& mstr) { ++count; });
+    int global_count = world.all_reduce_sum(count);
+    world.barrier();
+    YGM_ASSERT_RELEASE(global_count == 3);
```
end of above comment, the count will not always be 3

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+  //
+  // Test for_all (pair)
+  {
+    ygm::container::oversized_bag<std::pair<std::string, int>> pbag(world);
```
This highlights maybe a todo. we're mixing types in the binary_archive here. I'll have to discuss what the plan for this should be. Maybe appending the type of serialized data to the filename? idk. regardless a bbag.clear will be needed as cereal will try to read in existing binary data of a different type.

paul-rl replied:
I see what you mean, I think as long as it's modified for the bag to clear between tests, this should ensure that the files are deleted and the test works appropriately.

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+
+  // Test rebalance
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
```
needs a bbag.clear() this looks like a copy from line 102, is that intended?

paul-rl replied:
This was not intended, something that slipped through the cracks when merging, ty for pointing it out!

Daniel-DeArmas replied:
good catch

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+  //
+  // Test rebalance with non-standard rebalance sizes
+  {
+    ygm::container::oversized_bag<std::string> bbag(world);
```
will need a clear before this block

ryan-dozier commented on test/test_oversized_bag.cpp:
```diff
@@ -0,0 +1,222 @@
+  //
+  // Test output data after rebalance
+  {
+    ygm::container::oversized_bag<int> bbag(world);
```
clear before this block

ryan-dozier reviewed (commented):
Some issues to work through

Daniel-DeArmas commented on include/ygm/container/oversized_bag.hpp:
are these changes going to conflict with our oversized-bag-dev branch or are these changes based on those

paul-rl7 committed:
Addressed Ryan's comments in the merge requests, made tests isolated by clearing files at the end of tests

paul-rl commented:
Something important to note is that due to the fact that Ryan will end up overhauling a lot of the code base for his work, these tests might end up being being changed entirely. Functionally, for the scope of the project at hand, these are done. However, for the scope of Ryan's on work for his internship, there is still work to do.
